File: python/channel_section_localizations.py
#!/usr/bin/python

# Retrieve or set localized channel section metadata.
# See the Prerequisites section in the README file in this directory
# for more information about running this sample locally.

# Usage examples:
# python channel_sections_localizations.py --action=list --mine=True
# python channel_sections_localizations.py --action='<action>' --channel_section_id='<channel_section_id>' --default_language='<default_language>' --language='<language>' --title='<title>'

import argparse
import os
import re

import google.oauth2.credentials
import google_auth_oauthlib.flow
from googleapiclient.discovery import build
from googleapiclient.errors import HttpError
from google_auth_oauthlib.flow import InstalledAppFlow


# The CLIENT_SECRETS_FILE variable specifies the name of a file that contains

# the OAuth 2.0 information for this application, including its client_id and
# client_secret. You can acquire an OAuth 2.0 client ID and client secret from
# the {{ Google Cloud Console }} at
# {{ https://cloud.google.com/console }}.
# Please ensure that you have enabled the YouTube Data API for your project.
# For more information about using OAuth2 to access the YouTube Data API, see:
#   https://developers.google.com/youtube/v3/guides/authentication
# For more information about the client_secrets.json file format, see:
#   https://developers.google.com/api-client-library/python/guide/aaa_client_secrets
CLIENT_SECRETS_FILE = 'client_secret.json'

# This OAuth 2.0 access scope allows for full read/write access to the
# authenticated user's account.
SCOPES = ['https://www.googleapis.com/auth/youtube']
API_SERVICE_NAME = 'youtube'
API_VERSION = 'v3'

ACTIONS = ('get', 'list', 'set')

# Authorize the request and store authorization credentials.
def get_authenticated_service():
  flow = InstalledAppFlow.from_client_secrets_file(CLIENT_SECRETS_FILE, SCOPES)
  credentials = flow.run_console()
  return build(API_SERVICE_NAME, API_VERSION, credentials = credentials)

# Call the API's channelSections.update method to update an existing channel section's
# default language, and localized title in a specific language.
def set_channel_section_localization(youtube, args):

  # Retrieve the snippet and localizations for the channel section.
  results = youtube.channelSections().list(
    part='snippet,contentDetails,localizations',
    id=args.channel_section_id
  ).execute()

  channel_section = results['items'][0]

  # If the language argument is set, set the localized title for that language.
  # The "title" argument has a default value to make the script simpler to run
  # as a demo. But in an actual app, you would likely want to set its value.
  if args.language and args.language != '':
    if 'localizations' not in channel_section:
      channel_section['localizations'] = {}

    channel_section['localizations'][args.language] = {
      'title': args.title,
    }

  # If the default language is set AND there is localized metadata for that
  # language, set the channel section's title and description to match the
  # localized title and description for the newly set default language.
  if args.default_language:
    channel_section['snippet']['defaultLanguage'] = args.default_language
    if args.default_language in channel_section['localizations']:
      channel_section['snippet']['title'] = (
          channel_section['localizations'][args.default_language]['title'])

  update_result = youtube.channelSections().update(
    part='snippet,contentDetails,localizations',
    body=channel_section
  ).execute()

  if args.language:
    for language in update_result['localizations']:
      # Languages with locales, like "pt-br" are returned as pt-BR in metadata.
      # This ensures that the language specified when running the script can be
      # matched to the language returned in the metadata.
      if language.lower() == args.language.lower():
        localization = update_result['localizations'][args.language]
        print ('Updated channel section %s localized title to %s in \'%s\'.' %
               (args.channel_section_id, localization['title'], args.language))
        break

  if (args.default_language and
      args.default_language == update_result['snippet']['defaultLanguage']):
    print 'Updated default language to %s.' % args.default_language

# Call the API's channelSections.list method to retrieve an existing channel section localization.
# If the localized text is not available in the requested language,
# this method will return text in the default language.
def get_channel_section_localization(youtube, args):
  results = youtube.channelSections().list(
    part='snippet',
    id=args.channel_section_id,
    hl=args.language
  ).execute()

  # The localized object contains localized text if the hl parameter specified
  # a language for which localized text is available. Otherwise, the localized
  # object will contain metadata in the default language.
  localized = results['items'][0]['snippet']['localized']

  print('The channel section\'s title is \'%s\' in language \'%s\'.' %
        (localized['title'], language))

# Call the API's channelSections.list method to list all existing localizations
# for the channel section.
def list_channel_section_localizations(youtube, args):
  results = youtube.channelSections().list(
    part='snippet,localizations',
    id=args.channel_section_id
  ).execute()

  localizations = results['items'][0]['localizations']

  for language, localization in localizations.iteritems():
    print('The channel section title is \'%s\' in language \'%s\'.' %
          (localization['title'], language))

# Call the API's channelSections.list method to list localizations for all
# channel sections in the authorizing user\'s channel. This function might
# be called as a way of identifying the ID of the section you actually want
# to update.
def list_my_channel_section_localizations(youtube, args):
  results = youtube.channelSections().list(
    part='snippet,localizations',
    mine=True,
  ).execute()

  print(results)

  for i in range(0, len(results['items'])):
    item = results['items'][i]
    print str(item['snippet']['position']) + ':'
    print '    ID: ' + item['id']
    print '    Type: ' + item['snippet']['type']
    if ('title' in item['snippet'] and item['snippet']['title']):
      print '    Title: ' + str(item['snippet']['title'])

    if 'localizations' in results['items'][i]:
      localizations = results['items'][i]['localizations']
      print('    Localized titles by language:')
      for language, localization in localizations.iteritems():
        print('       ' + language + ': ' + localization['title'])
    else:
      print('    No localizations. :(\n')

  #for language, localization in localizations.iteritems():
  #  print('The channel section title is \'%s\' in language \'%s\'.' %
  #        (localization['title'], language))


if __name__ == '__main__':
  parser = argparse.ArgumentParser()
  # The 'action' option specifies the action to be processed.
  parser.add_argument('--action', choices=ACTIONS, required=True,
      help='The type of operation. Supported values are: "get", "list", "set"')
  # The ID of the channel section for which data is being retrieved or updated.
  parser.add_argument('--channel_section_id',
      help='ID of channel section for which localized data is being ' +
           'retrieved or updated.')
  # The language of the channel section's default metadata.
  parser.add_argument('--default_language',
      help='Default language of the channel section to update.')
  # The language of the localization that is being processed.
  parser.add_argument('--language', help='Language of the localized metadata.')
  # The localized channel section title for the specified language.
  parser.add_argument('--title',
      help='Localized title of the channel section to be set.',
      default='Localized Title')
  # The language of the channel section's default metadata.
  parser.add_argument('--mine', type=bool, default=False,
      help='List localizations for all of my channel sections.')

  args = parser.parse_args()

  if not args.channel_section_id and not args.mine:
    exit('You must either specify a channel section ID using the ' +
         '--channel_section_id argument or retrieve localizations ' +
         'for all of your channel sections by setting the --mine ' +
         'argument to True.')

  youtube = get_authenticated_service()
  try:
    if args.action == 'set':
      set_channel_section_localization(youtube, args)
    elif args.action == 'get':
      get_channel_section_localization(youtube, args)
    elif args.action == 'list':
      if args.mine:
        list_my_channel_section_localizations(youtube, args)
      else:
        list_channel_section_localizations(youtube, args)
    else:
      exit('Please specify a valid action using the --action= parameter.')
  except HttpError, e:
    print 'An HTTP error %d occurred:\n%s' % (e.resp.status, e.content)

File: php/channel_section_localizations.php
<?php

/**
 * This sample sets and retrieves localized metadata for a channel section by:
 *
 * 1. Updating language of the default metadata and setting localized metadata
 *   for a channel section via "channelSections.update" method.
 * 2. Getting the localized metadata for a channel section in a selected language using the
 *   "channelSections.list" method and setting the "hl" parameter.
 * 3. Listing the localized metadata for a channel section using the "channelSections.list" method
 *   and including "localizations" in the "part" parameter.
 *
 * @author Ibrahim Ulukaya
 */

/**
 * Library Requirements
 *
 * 1. Install composer (https://getcomposer.org)
 * 2. On the command line, change to this directory (api-samples/php)
 * 3. Require the google/apiclient library
 *    $ composer require google/apiclient:~2.0
 */
if (!file_exists(__DIR__ . '/vendor/autoload.php')) {
  throw new \Exception('please run "composer require google/apiclient:~2.0" in "' . __DIR__ .'"');
}

require_once __DIR__ . '/vendor/autoload.php';
session_start();

$htmlBody = <<<END
<form method="GET">
  <div>
    Action:
    <select id="action" name="action">
      <option value="set">Set Localization - Fill in: channel section ID, default language, language, title</option>
      <option value="get">Get Localization- Fill in: channel section ID, language</option>
      <option value="list">List Localizations - Fill in: channel section ID, language</option>
    </select>
  </div>
  <br>
  <div>
    Channel section ID: <input type="text" id="channelSectionId" name="channelSectionId" placeholder="Enter Channel Section ID">
  </div>
  <br>
  <div>
    Default Language: <input type="text" id="defaultLanguage" name="defaultLanguage" placeholder="Enter Default Language">
  </div>
  <br>
  <div>
    Default Language: <input type="text" id="defaultLanguage" name="defaultLanguage" placeholder="Enter Default Language (BCP-47 language code)">
  </div>
  <br>
  <div>
    Language: <input type="text" id="language" name="language" placeholder="Enter Local Language (BCP-47 language code)">
  </div>
  <br>
  <input type="submit" value="GO!">
</form>
END;

/*
 * You can acquire an OAuth 2.0 client ID and client secret from the
 * {{ Google Cloud Console }} <{{ https://cloud.google.com/console }}>
 * For more information about using OAuth 2.0 to access Google APIs, please see:
 * <https://developers.google.com/youtube/v3/guides/authentication>
 * Please ensure that you have enabled the YouTube Data API for your project.
 */
$OAUTH2_CLIENT_ID = 'REPLACE_ME';
$OAUTH2_CLIENT_SECRET = 'REPLACE_ME';

$client = new Google_Client();
$client->setClientId($OAUTH2_CLIENT_ID);
$client->setClientSecret($OAUTH2_CLIENT_SECRET);

/*
 * This OAuth 2.0 access scope allows for full read/write access to the
 * authenticated user's account.
 */
$client->setScopes('https://www.googleapis.com/auth/youtube');
$redirect = filter_var('http://' . $_SERVER['HTTP_HOST'] . $_SERVER['PHP_SELF'],
    FILTER_SANITIZE_URL);
$client->setRedirectUri($redirect);

// Define an object that will be used to make all API requests.
$youtube = new Google_Service_YouTube($client);

// Check if an auth token exists for the required scopes
$tokenSessionKey = 'token-' . $client->prepareScopes();
if (isset($_GET['code'])) {
  if (strval($_SESSION['state']) !== strval($_GET['state'])) {
    die('The session state did not match.');
  }

  $client->authenticate($_GET['code']);
  $_SESSION[$tokenSessionKey] = $client->getAccessToken();
  header('Location: ' . $redirect);
}

if (isset($_SESSION[$tokenSessionKey])) {
  $client->setAccessToken($_SESSION[$tokenSessionKey]);
}

// Check to ensure that the access token was successfully acquired.
if ($client->getAccessToken()) {
  $htmlBody = '';
  // This code executes if the user enters an action in the form
  // and submits the form. Otherwise, the page displays the form above.
  if (isset($_GET['action'])) {
    $resource = $_GET['resource'];
    $channelSectionId = $_GET['channelSectionId'];
    $language = $_GET['language'];
    $defaultLanguage = $_GET['defaultLanguage'];
    $title = $_GET['title'];
    try {
      switch ($_GET['action']) {
        case 'set':
          setChannelSectionLocalization($youtube, $channelSectionId, $defaultLanguage,
              $language, $title, $htmlBody);
          break;
        case 'get':
          getChannelSectionLocalization($youtube, $channelSectionId, $language, $htmlBody);
          break;
        case 'list':
          listChannelSectionLocalizations($youtube, $channelSectionId, $htmlBody);
          break;
      }
    } catch (Google_Service_Exception $e) {
      $htmlBody .= sprintf('<p>A service error occurred: <code>%s</code></p>',
          htmlspecialchars($e->getMessage()));
    } catch (Google_Exception $e) {
      $htmlBody .= sprintf('<p>An client error occurred: <code>%s</code></p>',
          htmlspecialchars($e->getMessage()));
    }
  }
  $_SESSION[$tokenSessionKey] = $client->getAccessToken();
} elseif ($OAUTH2_CLIENT_ID == 'REPLACE_ME') {
  $htmlBody = <<<END
  <h3>Client Credentials Required</h3>
  <p>
    You need to set <code>\$OAUTH2_CLIENT_ID</code> and
    <code>\$OAUTH2_CLIENT_ID</code> before proceeding.
  <p>
END;
} else {
  // If the user hasn't authorized the app, initiate the OAuth flow
  $state = mt_rand();
  $client->setState($state);
  $_SESSION['state'] = $state;

  $authUrl = $client->createAuthUrl();
  $htmlBody = <<<END
  <h3>Authorization Required</h3>
  <p>You need to <a href="$authUrl">authorize access</a> before proceeding.<p>
END;
}


/**
 * Updates a channel section's default language and sets its localized metadata.
 *
 * @param Google_Service_YouTube $youtube YouTube service object.
 * @param string $channelSectionId The id parameter specifies the channel section ID
 * for the resource that is being updated.
 * @param string $defaultLanguage The language of the channel section's default metadata
 * @param string $language The language of the localized metadata
 * @param string $title The localized title to be set
 * @param $htmlBody - html body.
 */
function setChannelSectionLocalization(Google_Service_YouTube $youtube, $channelSectionId,
    $defaultLanguage, $language, $title, &$htmlBody) {
  // Call the YouTube Data API's channelSections.list method to retrieve channel sections.
  $channelSections = $youtube->channelSections->listChannelSections("snippet,localizations", array(
      'id' => $channelSectionId
  ));

  // If $channelSections is empty, the specified channel section was not found.
  if (empty($channelSections)) {
    $htmlBody .= sprintf('<h3>Can\'t find a channel section with channel section id: %s</h3>',
        $channelSectionId);
  } else {
    // Since the request specified a channel section ID, the response only
    // contains one channel section resource.
    $updateChannelSection = $channelSections[0];

    // Modify channel section's default language and localizations properties.
    // Ensure that a value is set for the resource's snippet.defaultLanguage property.
    $updateChannelSection['snippet']['defaultLanguage'] = $defaultLanguage;
    $localizations = $updateChannelSection['localizations'];

    if (is_null($localizations)) {
      $localizations = array();
    }
    $localizations[$language] = array('title' => $title);
    $updateChannelSection['localizations'] = $localizations;

    // Call the YouTube Data API's channelSections.update method to update an
    // existing channel section.
    $channelSectionUpdateResponse = $youtube->channels->update("snippet,localizations",
        $updateChannelSection);

    $htmlBody .= "<h2>Updated channel section</h2><ul>";
    $htmlBody .= sprintf('<li>(%s) default language: %s</li>', $channelSectionId,
        $channelSectionUpdateResponse['snippet']['defaultLanguage']);
    $htmlBody .= sprintf('<li>title(%s): %s</li>', $language,
        $channelSectionUpdateResponse['localizations'][$language]['title']);
    $htmlBody .= '</ul>';
  }
}

/**
 * Returns localized metadata for a channel section in a selected language.
 * If the localized text is not available in the requested language,
 * this method will return text in the default language.
 *
 * @param Google_Service_YouTube $youtube YouTube service object.
 * @param string $channelSectionId The channelSectionId parameter instructs the API to return the
 * localized metadata for the channel section specified by the channel section id.
 * @param string language The language of the localized metadata.
 * @param $htmlBody - html body.
 */
function getChannelSectionLocalization(Google_Service_YouTube $youtube, $channelSectionId,
    $language, &$htmlBody) {
  // Call the YouTube Data API's channelSections.list method to retrieve channel sections.
  $channelSections = $youtube->channelSections->listChannelSections("snippet", array(
      'id' => $channelSectionId,
      'hl' => $language
  ));

  // If $channelSections is empty, the specified channel section was not found.
  if (empty($channelSections)) {
    $htmlBody .= sprintf('<h3>Can\'t find a channel section with channel section id: %s</h3>',
        $channelSectionId);
  } else {
    // Since the request specified a channel section ID, the response only
    // contains one channel section resource.
    $localized = $channelSections[0]["snippet"]["localized"];

    $htmlBody .= "<h3>Channel Section</h3><ul>";
    $htmlBody .= sprintf('<li>title(%s): %s</li>', $language, $localized['title']);
    $htmlBody .= '</ul>';
  }
}

/**
 * Returns a list of localized metadata for a channel section.
 *
 * @param Google_Service_YouTube $youtube YouTube service object.
 * @param string $channelId The channelSectionId parameter instructs the API to return the
 * localized metadata for the channel section specified by the channel section id.
 * @param $htmlBody - html body.
 */
function listChannelSectionLocalizations(Google_Service_YouTube $youtube,
    $channelSectionId, &$htmlBody) {
  // Call the YouTube Data API's channelSections.list method to retrieve channel sections.
  $channelSections = $youtube->channelSections->listChannelSections("snippet", array(
      'id' => $channelSectionId
  ));

  // If $channelSections is empty, the specified channel section was not found.
  if (empty($channelSections)) {
    $htmlBody .= sprintf('<h3>Can\'t find a channel section with channel section id: %s</h3>',
        $channelSectionId);
  } else {
    // Since the request specified a channel section ID, the response only
    // contains one channel section resource.
    $localizations = $channelSections[0]["localizations"];

    $htmlBody .= "<h3>Channel Section</h3><ul>";
    foreach ($localizations as $language => $localization) {
      $htmlBody .= sprintf('<li>title(%s): %s</li>', $language, $localization['title']);
    }
    $htmlBody .= '</ul>';
  }
}
?>

<!doctype html>
<html>
<head>
<title>Set and retrieve localized metadata for a channel section</title>
</head>
<body>
  <?=$htmlBody?>
</body>
</html>
